---
title: "R Notebook"
---




```{r}
library(tidyverse)
library(glue)
library(SingleCellExperiment)
source("util.R")
```



```{r}
if(! file.exists("../benchmark/output/glioblastoma/fit_hvg10k-small.qs")){
  
  sce <- qs::qread("../benchmark/data/glioblastoma/glioblastoma_annotated_sce.qs")
  colnames(sce) <- paste0("cell_", seq_len(ncol(sce)))
  hvg <- order(-rowVars(logcounts(sce)))
  sce <- sce[hvg[1:10000],]
  
  set.seed(1)
  # fit <- lemur::lemur(sce, design = ~ condition + patient_id, n_embedding = 50, test_fraction = 0.7)
  fit <- lemur::lemur(sce, design = ~ condition + patient_id, n_embedding = 60, test_fraction = 0.4)
  reducedDim(fit, "fit_umap") <- uwot::umap(t(fit$embedding))
  fit <- lemur::align_harmony(fit)
  reducedDim(fit, "fit_al_umap") <- uwot::umap(t(fit$embedding))
  fit <- lemur::test_de(fit, contrast = cond(condition = "panobinostat") - cond(condition = "ctrl"), new_assay_name = "DE_panobinostat")
  fit <- lemur::test_de(fit, contrast = cond(condition = "etoposide") - cond(condition = "ctrl"), new_assay_name = "DE_etoposide")
  
  qs::qsave(fit, "../benchmark/data/glioblastoma/fit_hvg10k.qs")
  set.seed(1)
  nei_pan <- lemur::find_de_neighborhoods(fit, group_by = vars(patient_id, condition), 
                                          contrast = cond(condition = "panobinostat") - cond(condition = "ctrl"),
                                          de_mat = assay(fit, "DE_panobinostat"), test_method = "edgeR")
  qs::qsave(nei_pan, "../benchmark/data/glioblastoma/fit_hvg10k-nei_pan.qs")
  set.seed(1)
  nei_pan_tumor <- lemur::find_de_neighborhoods(fit[,fit$colData$cell_type %in% c("Tumor")], group_by = vars(patient_id, condition), 
                                          contrast = cond(condition = "panobinostat") - cond(condition = "ctrl"),
                                          de_mat = assay(fit, "DE_panobinostat")[,fit$colData$cell_type %in% c("Tumor")], test_method = "edgeR")
  qs::qsave(nei_pan_tumor, "../benchmark/data/glioblastoma/fit_hvg10k-nei_pan_tumor.qs")
  
  
  psce_global <- glmGamPoi::pseudobulk(fit, group_by = vars(patient_id, condition), n = n())
  glm_fit <- lemur:::edger_fit(counts(psce_global), design = ~ patient_id + condition, col_data = colData(psce_global),
                             offset = log(glmGamPoi:::estimate_size_factors(counts(psce_global), method = "ratio")))
  glm_de <- lemur:::edger_test_de(glm_fit, contrast = c(0, 0, 0, 0, 0, 0, 0, 1))
  
  as_tibble(nei_pan) %>% arrange(did_pval)
  as_tibble(nei_pan_tumor) %>% arrange(pval)
  
  pan_affect_genes <- c("HIST3H2A", "CD84", "MBP", "CCL7")
  tumor_bin_genes <- c("PDIA4", "HERPUD1")
  cell_type_markers <- c("CD14", "TYROBP", "PLP1", "MAG", "TRBC1", "TRBC2", "GAP43", "B2M") 
  genes_of_interest <- rowData(fit) %>%
    as_tibble() %>%
    dplyr::select(gid, gene) %>%
    filter(gene %in% c(pan_affect_genes, tumor_bin_genes, cell_type_markers))
  
  fit_small <- fit[genes_of_interest$gid,]
  nei_pan_small <- nei_pan %>% filter(name %in% filter(genes_of_interest, gene %in% pan_affect_genes)$gid)
  nei_pan_tumor_small <- nei_pan_tumor %>% filter(name %in% filter(genes_of_interest, gene %in% tumor_bin_genes)$gid)
  # PDIA4 ENSG00000155660 
  # HERPUD1 ENSG00000051108
  inside_herpud1_nei <- colnames(fit[,fit$colData$cell_type == "Tumor"]) %in% filter(nei_pan_tumor_small, name == "ENSG00000051108")$neighborhood[[1]]
  psce_herpud1 <- glmGamPoi::pseudobulk(fit[,fit$colData$cell_type == "Tumor"], group_by = vars(patient_id, condition, inside = inside_herpud1_nei), n = n())
  
  qs::qsave(fit_small, "../benchmark/output/glioblastoma/fit_hvg10k-small.qs")
  qs::qsave(nei_pan_small, "../benchmark//output/glioblastoma/fit_hvg10k-nei_pan-small.qs")
  qs::qsave(nei_pan_tumor_small, "../benchmark//output/glioblastoma/fit_hvg10k-nei_pan_tumor-small.qs")
  qs::qsave(psce_herpud1, "../benchmark/output/glioblastoma/fit_hvg10k-herpud1_psce.qs")
  qs::qsave(rowData(fit), "../benchmark/output/glioblastoma/fit_hvg10-rowdata.qs")
  qs::qsave(nei_pan %>% dplyr::select(-neighborhood), "../benchmark/output/glioblastoma/fit_hvg10-nei_pan-without_neighborhood.qs")
}

```




```{r}
fit_small <- qs::qread("../benchmark/output/glioblastoma/fit_hvg10k-small.qs")
nei_pan_small <- qs::qread("../benchmark/output/glioblastoma/fit_hvg10k-nei_pan-small.qs")
nei_pan_mod <- qs::qread("../benchmark/output/glioblastoma/fit_hvg10-nei_pan-without_neighborhood.qs")
nei_pan_tumor_small <- qs::qread("../benchmark/output/glioblastoma/fit_hvg10k-nei_pan_tumor-small.qs")
psce_herpud1 <- qs::qread("../benchmark/output/glioblastoma/fit_hvg10k-herpud1_psce.qs")
full_row_data <- qs::qread("../benchmark/output/glioblastoma/fit_hvg10-rowdata.qs")
```

```{r}
cell_type_colors <- structure(ggsci::pal_npg()(4), names = c("Tumor", "T cell", "Oligodendrocytes", "Myeloid"))
ggplot(enframe(cell_type_colors), aes(x = name, y = 1)) +
    geom_tile(aes(fill = value)) +
    scale_fill_identity()
```


```{r}
# RColorBrewer::display.brewer.all()

condition_colors<- structure(RColorBrewer::brewer.pal(n = 3, name = "Set2"), names = c("ctrl", "panobinostat", "etoposide"))
ggplot(enframe(condition_colors), aes(x = name, y = 1)) +
    geom_tile(aes(fill = value)) +
    scale_fill_identity()
```

```{r, paged.print = FALSE}
# RColorBrewer::display.brewer.all()

pat_cond_color_df <- tibble(patient_id = c("PW029", "PW030", "PW032", "PW034", "PW036", "PW040")) %>%
  # mutate(ctrl_color = RColorBrewer::brewer.pal(n = 6, name = "Set2")) %>%
  mutate(ctrl_color = colorspace::qualitative_hcl(6, palette = "Set2") ) %>%
  mutate(panobinostat_color = colorspace::darken(ctrl_color, amount = 0.2),
         etoposide_color = colorspace::lighten(ctrl_color, amount = 0.4)) %>%
  pivot_longer(ends_with("color"), names_sep = "_", names_to = c("condition", ".value")) %>%
  mutate(condition = factor(condition, levels = c("etoposide", "ctrl", "panobinostat"))) 

pat_cond_color_df %>%
  ggplot(aes(x = patient_id, y = condition)) +
    geom_tile(aes(fill = color)) +
    scale_fill_identity()

pat_cond_color <- pat_cond_color_df %>%
  transmute(pat_cond = paste0(patient_id, "-", condition), color = color) %>%
  deframe()

```




```{r, paged.print=FALSE}
sel_genes <- c("CD14", "TYROBP", "PLP1", "MAG", "TRBC1", "TRBC2", "GAP43", "B2M") 
ct_marker_genes <- tibble(gene = sel_genes, cell_type_marker = rep(c("Myeloid", "Oligodendrocytes", "T cell", "Tumor"), each = 2))

stopifnot(all(sel_genes %in% rowData(fit_small)$gene))
gene_ids <- deframe(as_tibble(rowData(fit_small)[c("gene", "gid")]))[sel_genes]
expr_mat <- counts(fit_small)[gene_ids, ]
cell_type_dot_plot <- as_tibble(as.matrix(t(expr_mat))) %>%
  mutate(cell_type = fit_small$colData$cell_type) %>%
  pivot_longer(-cell_type, names_to = "gene_id", values_to = "expr") %>%
  left_join(enframe(gene_ids,  name = "gene_name",value = "gene_id"), by = "gene_id") %>%
  summarize(mean_expr = mean(expr, trim = 0),
            frac_expr = mean(expr != 0),
            .by = c(cell_type, gene_id, gene_name)) %>%
  mutate(gene_name = fct_rev(factor(gene_name, levels = sel_genes))) %>%
  mutate(cell_type_short = case_when(
    cell_type == "Myeloid" ~  "Myel.",
    cell_type == "Oligodendrocytes" ~  "Oligo.",
    TRUE ~ cell_type
  )) %>%
  left_join(ct_marker_genes, by = c("gene_name" = "gene")) %>%
  ggplot() +
    geom_point(aes(x = cell_type_short, y = gene_name, color = log2(mean_expr + 1), size = frac_expr), stroke = 0) +
    geom_rect(data = ct_marker_genes, aes(xmin = 0.7, xmax = 0.75, ymin = -Inf, ymax=Inf, fill = cell_type_marker), show.legend=FALSE) +
    facet_grid(rows = vars(cell_type_marker), scales = "free_y") +
    colorspace::scale_color_continuous_sequential(limits = c(0, NA), breaks = c(0, 2, 4)) +
    scale_fill_manual(values = cell_type_colors) +
    scale_size_binned(range = c(0.01, 3), breaks = c(0, 0.25, 0.75, 1), limits = c(0, 1), labels = c(0, "", "", 1)) +
    scale_x_discrete(expand = expansion(add = 0.3)) +
    scale_y_discrete(expand = expansion(add = 0.4)) +
    coord_cartesian(clip = "off") +
    guides(color = guide_colorbar(barheight = unit(1, "mm"), barwidth = unit(5, "mm")),
           size = guide_bins()) +
    theme(axis.title = element_blank(), legend.position = c(-0.2,-0.2), legend.direction = "horizontal",
          legend.box = "horizontal", legend.text = element_text(size = font_size_tiny),
          strip.background = element_blank(), strip.text = element_blank(), panel.spacing.y = unit(3, "mm")) +
    labs(size = "$\\text{frac}[y \\neq 0]$", color = "$\\log_2(\\bar{y} + 1)$")

cell_type_dot_plot
```


```{r, paged.print=FALSE}
sel_genes <- c("CD14", "PLP1", "GAP43", "HIST3H2A")

top_dirs <- deframe(full_row_data[,c("gene", "gid")])[sel_genes]

ctrl_proj <- lemur:::grassmann_map(lemur:::sum_tangent_vectors(fit_small$coefficients, c(1, 0, 0, rep(0, times = 5))), fit_small$base_point)[,1:2]
rownames(ctrl_proj) <- full_row_data$gid
ctrl_proj_df <- as_tibble(ctrl_proj[top_dirs,], rownames = "gid") %>%
  left_join(as_tibble(full_row_data))

pan_proj <- lemur:::grassmann_map(lemur:::sum_tangent_vectors(fit_small$coefficients, c(1, 0, 1, rep(0, times = 5))), fit_small$base_point)[,1:2]
rownames(pan_proj) <- full_row_data$gid
pan_proj_df <- as_tibble(pan_proj[top_dirs,], rownames = "gid") %>%
  left_join(as_tibble(full_row_data))

proj_df <- bind_rows(mutate(ctrl_proj_df, condition = "ctrl"),
                     mutate(pan_proj_df, condition = "panobinostat")) %>%
  # filter(gene != "HIST3H2A") %>%
  mutate(V1 = ifelse(gene == "HIST3H2A", V1 * 20, V1),
         V2 = ifelse(gene == "HIST3H2A", V2 * 20, V2))

manual_text_pos <- tribble(
      ~gene,      ~condition,   ~V1,  ~V2, ~hjust, ~vjust,
      "CD14",         "ctrl",   -17,   -4,     1,       0,
   #  "CD14", "panobinostat",   -20,   -9,     1,     0.3,
      "PLP1",         "ctrl",     1,   18,   0.5,       0,
   #  "PLP1", "panobinostat",     1,   17,   0.5,       0,
     "GAP43",         "ctrl",    19,  -13,     0,       1,
   # "GAP43", "panobinostat",    20,   -9,     0,       1,
  "HIST3H2A",         "ctrl",   -35,    4,     0,       0,
# "HIST3H2A", "panobinostat",   -14,  -15,   0.2,       1,
)

# eto_proj <- lemur:::grassmann_map(lemur:::sum_tangent_vectors(fit_small$coefficients, c(1, 1, 0, rep(0, times = 5))), fit_small$base_point)[,1:2]
# rownames(eto_proj) <- full_row_data$gid
# eto_proj_df <- as_tibble(eto_proj[top_dirs,] * c(rep(200, 3), 2000), rownames = "gid") %>%
#   left_join(as_tibble(full_row_data))

pan_tang <- lemur:::sum_tangent_vectors(fit_small$coefficients, c(1, 0, 1, rep(0, times = 5)))[,1:2] * 300
rownames(pan_tang) <- full_row_data$gid


mc_biplot <- as_tibble(colData(fit_small)) %>%
  mutate(emb = t(fit_small$embedding[1:2,])) %>%
  ggplot(aes(x = -emb[,1], y = -emb[,2])) +
    ggrastr::rasterize(geom_point(aes(color = cell_type), size = 0.01, stroke = 0, show.legend = FALSE), dpi = 600) +
    scale_color_manual(values = cell_type_colors) +
    ggnewscale::new_scale_colour() +
    geom_segment(data = proj_df , aes(x = 0, y = 0, xend = V1 * -200, yend = V2 * -200, color = condition), linewidth = 1.2, show.legend = FALSE) +
    annotate("segment", x = -35, y = 3.3, xend = 0.115 * -200, yend = 0.0787 * -200, color = "#222222", linewidth = 0.4) +
    annotate("segment", x = -14, y = 3.3, xend = 0.0193 * -200, yend = -0.0135 * -200, color = "#222222", linewidth = 0.4) +
    scale_color_manual(values = condition_colors) +
    geom_text(data = manual_text_pos, aes(x = V1, y = V2, hjust = hjust, vjust = vjust, label = gene), size = font_size_small / .pt) +
    coord_fixed(ylim = c(-20, 25), xlim = c(-40, 20), clip = "off") +
    labs(x = "$Z_1$", y ="$Z_2$")
    
mc_biplot
```



Make UMAPs

```{r}
make_dim_pl <- function(red_dim, color_by = NULL, axis_label = ""){
  as_tibble(colData(fit_small)) %>%
    mutate(umap = reducedDim(fit_small, red_dim)) %>%
    sample_frac(n = 1) %>%
    ggplot(aes(x = umap[,1], y = umap[,2])) +
      ggrastr::rasterize(geom_point(aes(color = {{color_by}}), size = 0.01, stroke = 0, show.legend = FALSE), dpi = 600) +
      small_axis(label = axis_label)
}



umap_orig_ct_pl <- make_dim_pl("UMAP", color_by = cell_type, axis_label = "UMAP") +
  scale_color_manual(values = cell_type_colors)
umap_orig_pc_pl <- make_dim_pl("UMAP", color_by = condition, axis_label = "UMAP") +
  geom_text(data = . %>% filter(cell_type == "Tumor") %>% summarize(umap = matrix(colMedians(umap), nrow = 1), .by = patient_id),
            aes(label = patient_id), size = font_size_tiny / .pt) +
  scale_color_manual(values = condition_colors)

umap_fit_ct_pl <- make_dim_pl("fit_umap", color_by = cell_type, axis_label = "UMAP") +
  scale_color_manual(values = cell_type_colors)
umap_fit_pc_pl <- make_dim_pl("fit_umap", color_by = condition, axis_label = "UMAP") +
  scale_color_manual(values = condition_colors)

umap_fital_ct_pl <- make_dim_pl("fit_al_umap", color_by = cell_type, axis_label = "UMAP") +
  scale_color_manual(values = cell_type_colors)
umap_fital_pc_pl <-make_dim_pl("fit_al_umap", color_by = condition, axis_label = "UMAP") +
  scale_color_manual(values = condition_colors)

umap_orig_ct_pl
umap_orig_pc_pl
```




```{r, paged.print=FALSE}
scale_sym <- function(x, ignore_extreme = 0){
  abs_max <- max(abs(quantile(x, c(ignore_extreme, 1-ignore_extreme))))
  x / abs_max
}

umap_fit <- reducedDim(fit_small, "fit_al_umap")
pan_genes_order <- c("HIST3H2A", "CCL7")

is_inside <- tibble(gid = rowData(fit_small)$gid, gene = rowData(fit_small)$gene, cell_id = list(colnames(fit_small))) %>%
  filter(gene %in% pan_genes_order) %>%
  left_join(dplyr::select(nei_pan_small, name, neighborhood), by = c("gid"= "name")) %>%
  mutate(inside = map2(cell_id, neighborhood, \(ref, nei_cells)ref  %in% nei_cells)) %>%
  dplyr::select(-neighborhood) %>%
  unnest(c(inside, cell_id)) %>%
  mutate(gene = factor(gene, levels = pan_genes_order))


de_regions_maps <- nei_pan_small %>%
  unnest(neighborhood) %>%
  left_join(tibble(umap = umap_fit, neighborhood = rownames(umap_fit)), by = "neighborhood") %>%
  left_join(as_tibble(full_row_data), by = c("name" = "gid")) %>%
  filter(gene %in% pan_genes_order) %>%
  mutate(gene = factor(gene, levels = pan_genes_order)) 
  

de_regions_maps2 <- nei_pan_tumor_small %>%
  unnest(neighborhood) %>%
  left_join(tibble(umap = umap_fit, neighborhood = rownames(umap_fit)), by = "neighborhood") %>%
  left_join(as_tibble(full_row_data), by = c("name" = "gid"))


de_plot_data <- as_tibble(colData(fit_small), rownames = "cell_id") %>%
  mutate(umap = umap_fit) %>%
  mutate(de = as_tibble(t(assay(fit_small, "DE_panobinostat")))) %>%
  unnest(de, names_sep = "-") %>%
  pivot_longer(starts_with("de-"), names_sep = "-", values_to = "de", names_to = c(NA, "gid")) %>%
  inner_join(as_tibble(full_row_data), by = "gid") %>%
  filter(gene %in% pan_genes_order) %>%
  mutate(gene = factor(gene, levels = pan_genes_order)) %>%
  left_join(is_inside, by = c("gid", "gene", "cell_id")) %>%
  mutate(inside = ifelse(inside, "in", "out")) %>%
  mutate(inside = factor(inside, levels = c("in", "out")))  %>%
  group_by(gene) %>%
  mutate(de = scale_sym(de, ignore_extreme = 0.05)) %>%
  sample_frac(size = 1) 

de_plots <- de_plot_data %>%
  group_by(gene) %>%
  group_map(\(data, key){
    ggplot(data, aes(x = umap[,1], y = umap[,2])) +
      ggrastr::rasterise(geom_point(aes(color = de), size = 0.05, stroke = 0, show.legend = FALSE), dpi = 600) +
      # geom_density2d(data = de_regions_maps %>% filter(gene == key[[1]]) %>% mutate(inside = "in"),  breaks = 0.2, linewidth = 0.75,
      #                contour_var = "ndensity", color = "black",
      #                h = apply(umap_fit, 2, MASS::bandwidth.nrd) * 0.9,
      #                show.legend = FALSE) +
      scale_colour_gradient2_rev(limits = c(-1, 1), oob = scales::squish, breaks = c(-1, 0, 1)) +
      facet_wrap(vars(inside), nrow = 1, labeller = as_labeller(c("in" = "Inside Neighborhood", "out" = "Outside Neighborhood")))  +
      scale_x_continuous(limits = range(umap_fit[,1]) * 1.1) +
      scale_y_continuous(limits = range(umap_fit[,2]) * 1.1) +
      small_axis(label = "", fontsize = font_size_tiny, xlim = range(umap_fit[,1]), ylim = range(umap_fit[,2]), arrow_length = 3) +
      theme(panel.spacing.x = unit(7, "mm"), plot.subtitle = element_text(margin = margin(2, 0, 0, 0, unit = "pt"))) +
      labs(subtitle = key[[1]])
  })

de_plots
```


```{r, paged.print=FALSE}
expr_plot_data <- as_tibble(colData(fit_small), rownames = "cell_id") %>%
  mutate(umap = umap_fit) %>%
  mutate(expr = as_tibble(as.matrix(t(assay(fit_small, "logcounts"))))) %>%
  unnest(expr, names_sep = "-") %>%
  pivot_longer(starts_with("expr-"), names_sep = "-", values_to = "expr", names_to = c(NA, "gid")) %>%
  inner_join(as_tibble(full_row_data), by = "gid") %>%
  sample_frac(size = 1)  %>%
  filter(gid %in% nei_pan_small$name | gid %in% nei_pan_tumor_small$name)



expr_plots <- expr_plot_data %>%
  filter(condition != "etoposide") %>%
  mutate(gene = factor(gene, levels = c(pan_genes_order, "HERPUD1"))) %>%
  filter(! is.na(gene)) %>%
  group_by(gene) %>%
  group_map(\(data, key){
    data %>% mutate(gene = key[[1]]) %>%
      ggplot(aes(x = umap[,1], y = umap[,2])) +
      ggrastr::rasterise(geom_point(aes(color = expr), size = 0.05, stroke = 0), dpi = 300) +
      geom_density2d(data = filter(rbind(de_regions_maps %>% mutate(gene = as.character(gene)), de_regions_maps2), gene == key[[1]]), breaks = 0.2, linewidth = 0.5,
                     contour_var = "ndensity", color = "black",
                     h = apply(umap_fit, 2, MASS::bandwidth.nrd) * 0.9,
                     show.legend = FALSE) +
      colorspace::scale_color_continuous_sequential(limits = c(0, quantile(data$expr, c(0.98))), oob = scales::oob_squish, palette = "Purples 2", n.breaks = 3) +
      facet_grid(vars(condition), vars(gene))  +
      scale_x_continuous(limits = range(umap_fit[,1]) * 1.1) +
      scale_y_continuous(limits = range(umap_fit[,2]) * 1.1) +
      small_axis(label = "", fontsize = font_size_tiny, xlim = range(umap_fit[,1]), ylim = range(umap_fit[,2]), arrow_length = 3) +
      guides(color = guide_colorbar(barheight = unit(2, "mm"), title = "")) +
      theme(panel.spacing.x = unit(5, "mm"), legend.position =  "bottom") +
      labs(color = "")
  }) %>%
  cowplot::plot_grid(plotlist = ., nrow = 1)



expr_plots
```


```{r, paged.print=FALSE}
plot_assemble(add_text("Expression pattern of selected genes under panobinostat treatment", x = 1.5, y = 1.5, vjust = 1, fontsize = font_size, fontface = "bold"),
              add_plot(expr_plots, x = 0, y = 5, width = 170, height = 80),
              width = 170, height = 85, units = "mm", show_grid_lines = FALSE, latex_support = TRUE,
              filename = "../plots/suppl_gene_expr_patterns.pdf")
```


```{r}
pan_genes_ids <- deframe(full_row_data[, c("gene", "gid")])[pan_genes_order]

mask <- matrix(NA, nrow = length(pan_genes_ids), ncol = ncol(fit_small), 
               dimnames = list(pan_genes_ids, colnames(fit_small)))
mask2 <- matrix(1, nrow = length(pan_genes_ids), ncol = ncol(fit_small), 
               dimnames = list(pan_genes_ids, colnames(fit_small)))
for(id in pan_genes_ids){
  mask[id, filter(nei_pan_small, name == id)$neighborhood[[1]]] <- 1
  mask2[id, filter(nei_pan_small, name == id)$neighborhood[[1]]] <- NA
}

psce <- glmGamPoi::pseudobulk(SingleCellExperiment(list(inside = as.matrix(logcounts(fit_small)[pan_genes_ids,] * mask))),
                      group_by = vars(patient_id, condition, pat_cond), n = n(),
                      aggregation_functions = list(.default = \(...) matrixStats::rowMeans2(..., na.rm = TRUE)),
                      col_data = as.data.frame(colData(fit_small)))

comparison_pl <- as_tibble(colData(psce)) %>%
  mutate(expr_inside = as_tibble(t(assay(psce, "inside")))) %>%
  unpack(starts_with("expr"), names_sep = "-") %>%
  pivot_longer(starts_with("expr"), names_sep = "[-_]", names_to = c(".value", "origin", "gid")) %>%
  left_join(as_tibble(full_row_data)) %>%
  filter(condition != "etoposide") %>%
  mutate(condition_short = case_when(
    condition == "ctrl" ~ "ctrl",
    condition == "panobinostat" ~ "pano.",
  )) %>%
  mutate(gene = factor(gene, levels = pan_genes_order)) %>%
  group_by(gene) %>%
  group_map(\(data, key){
    ggplot(data, aes(x = condition_short, y = expr)) +
      geom_point(aes(group = condition, color = condition), position = position_dodge(width = 0.7), size = 0.6, show.legend = FALSE) +
      geom_line(aes(group = patient_id, x = condition_short, y = expr), color = "lightgrey", linewidth = 0.3) +
      scale_color_manual(values = condition_colors) +
      scale_y_continuous(limits = c(0, NA), expand = expansion(mult = c(0, 0.1)), n.breaks = 3) +
      coord_cartesian(clip = "off") +
      guides(x = guide_axis(angle = 90)) +
      theme(axis.title = element_blank(), strip.background = element_blank(), strip.text = element_blank(),
            axis.text = element_text(size = font_size_tiny),
            panel.spacing.x = unit(25, "mm")) 
  })

comparison_pl
```





```{r, paged.print=FALSE}

is_inside2 <- tibble(gid = rowData(fit_small)$gid, gene = rowData(fit_small)$gene, cell_id = list(colnames(fit_small))) %>%
  filter(gid %in% "ENSG00000051108") %>%
  left_join(dplyr::select(nei_pan_tumor_small, name, neighborhood), by = c("gid"= "name")) %>%
  mutate(inside = map2(cell_id, neighborhood, \(ref, nei_cells)ref  %in% nei_cells)) %>%
  dplyr::select(-neighborhood) %>%
  unnest(c(inside, cell_id))

de_plot_data2 <- as_tibble(colData(fit_small), rownames = "cell_id") %>%
  mutate(umap = umap_fit) %>%
  mutate(de = as_tibble(t(assay(fit_small, "DE_panobinostat")))) %>%
  unnest(de, names_sep = "-") %>%
  pivot_longer(starts_with("de-"), names_sep = "-", values_to = "de", names_to = c(NA, "gid")) %>%
  inner_join(as_tibble(full_row_data), by = "gid") %>%
  inner_join(is_inside2, by = c("gid", "cell_id", "gene")) %>%
  filter(gid == "ENSG00000051108") %>%
  group_by(gene) %>%
  mutate(de = scale_sym(de, ignore_extreme = 0.05)) %>%
  sample_frac(size = 1) 

de_plots2 <- de_plot_data2 %>%
  mutate(inside = ifelse(inside, "in", "out")) %>%
  mutate(inside = factor(inside, levels = c("in", "out")))  %>%
  mutate(de = ifelse(cell_type == "Tumor", de, NA)) %>%
  ggplot(aes(x = umap[,1], y = umap[,2])) +
    ggrastr::rasterise(geom_point(aes(color = de), size = 0.05, stroke = 0, show.legend = FALSE), dpi = 600) +
    # geom_density2d(data = de_regions_maps2 %>% mutate(inside = "in"),  breaks = 0.3, linewidth = 0.75,
    #                contour_var = "ndensity", color = "black",
    #                h = apply(umap_fit, 2, MASS::bandwidth.nrd) * 0.9,
    #                show.legend = FALSE) +
    scale_colour_gradient2_rev(limits = c(-1, 1), oob = scales::squish, breaks = c(-1, 0, 1), na.value = "#00000000") +
    facet_wrap(vars(inside), nrow = 1, labeller = as_labeller(c("in" = "Inside Neighborhood", "out" = "Outside Neighborhood")))  +
    scale_x_continuous(limits = range(umap_fit[,1]) * 1.1) +
    scale_y_continuous(limits = range(umap_fit[,2]) * 1.1) +
    small_axis(label = "", fontsize = font_size_tiny, xlim = range(umap_fit[,1]), ylim = range(umap_fit[,2]), arrow_length = 3) +
    theme(panel.spacing.x = unit(-2, "mm"))

de_plots2
```



```{r, paged.print = FALSE}
genes_of_interest <- c("ENSG00000051108")

mask <- matrix(NA, nrow = length(genes_of_interest), ncol = ncol(fit_small), 
               dimnames = list(genes_of_interest, colnames(fit_small)))
mask2 <- matrix(1, nrow = length(genes_of_interest), ncol = ncol(fit_small), 
               dimnames = list(genes_of_interest, colnames(fit_small)))
for(id in genes_of_interest){
  mask[id, filter(nei_pan_tumor_small, name == id)$neighborhood[[1]]] <- 1
  mask2[id, filter(nei_pan_tumor_small, name == id)$neighborhood[[1]]] <- NA
}

psce2 <- glmGamPoi::pseudobulk(SingleCellExperiment(list(inside = as.matrix(logcounts(fit_small)[genes_of_interest,] * mask),
                                                        outside = as.matrix(logcounts(fit_small)[genes_of_interest,] * mask2))),
                      group_by = vars(patient_id, condition, pat_cond), n = n(),
                      aggregation_functions = list(.default = \(...) matrixStats::rowMeans2(..., na.rm = TRUE)),
                      col_data = as.data.frame(colData(fit_small)))

comparison_data2 <- as_tibble(colData(psce2)) %>%
  mutate(expr_inside = as_tibble(t(assay(psce2, "inside"))),
         expr_outside = as_tibble(t(assay(psce2, "outside")))) %>%
  unpack(starts_with("expr"), names_sep = "-") %>%
  pivot_longer(starts_with("expr"), names_sep = "[-_]", names_to = c(".value", "origin", "gid")) %>%
  left_join(as_tibble(full_row_data)) %>%
  filter(condition != "etoposide") %>%
  mutate(condition_short = case_when(
    condition == "ctrl" ~ "ctrl",
    condition == "panobinostat" ~ "panob.",
  ))

diff_df <-  comparison_data2 %>% 
  summarize(expr = mean(expr), .by = c(condition, origin)) %>% 
  pivot_wider(names_from = condition, values_from = expr)

comparison_pl2 <-  comparison_data2 %>%
  filter(patient_id != "PW029") %>%
  ggplot(aes(x = condition_short, y = expr)) +
    # geom_boxplot(outlier.shape =  NA, width = 0.3) +
    geom_line(aes(group = patient_id, x = condition_short, y = expr), color = "lightgrey", linewidth = 0.3) +
    geom_point(aes(group = condition, color = condition), position = position_dodge(width = 0.7), size = 0.8, show.legend = FALSE) +
    geom_segment(data = diff_df, aes(x = 1.5, y = ctrl, xend = 1.5, yend = panobinostat), color = "red", linewidth = 0.8, arrow = arrow(type = "closed", length = unit(1, "mm"))) +
    facet_grid(cols = vars(origin), labeller = as_labeller(c("inside" = "Inside\nNeighborhood", "outside" = "Outside\nNeighborhood"))) +
    scale_color_manual(values = condition_colors) +
    scale_y_continuous(limits = c(0, NA), expand = expansion(mult = c(0, 0.05)), n.breaks = 3) +
    coord_cartesian(clip = "off") +
    # guides(x = guide_axis(angle = 90)) +
    theme(panel.spacing.x = unit(1, "mm"), axis.title.x = element_blank()) +
    labs(y = "pseudobulked expr.")

diff_pl <- diff_df %>% 
  ggplot(aes(x = origin, xend = origin, y = 0, yend = panobinostat - ctrl)) +
    geom_hline(yintercept = 0) +
    geom_segment(color = "red", linewidth = 0.8, arrow = arrow(type = "closed", length = unit(1, "mm"))) +
    ggsignif::geom_signif(comparison = list(c("inside", "outside")), annotations = "**", textsize = font_size_small / .pt, y_position = 1.3) +
    scale_y_continuous(limits = c(0, max(comparison_data2$expr)) - 1, expand = expansion(mult = c(0.05, 0.05)), n.breaks = 3) +
    labs(subtitle = "$ \\text{Diff. in diff.}$\n($\\Delta_{\\text{in}} - \\Delta_{\\text{out}}$)",
         y = "$\\Delta$") +
    theme(axis.title.x = element_blank(), plot.subtitle = element_text(hjust = 0.5)) 

comparison_pl2
diff_pl
```


```{r, paged.print=FALSE}
glm_fit <- glmGamPoi::glm_gp(psce_herpud1[,psce_herpud1$condition == "ctrl"], design = ~ patient_id + inside, size_factors = "ratio", verbose = TRUE)
glm_de_res <- glmGamPoi::test_de(glm_fit, cond(inside = TRUE) - cond(inside = FALSE))
glm_fit2 <- lemur:::edger_fit(counts(psce_herpud1[,psce_herpud1$condition == "ctrl"]), design = ~ patient_id + inside, col_data = colData(psce_herpud1[,psce_herpud1$condition == "ctrl"]),
                             offset = log(glmGamPoi:::estimate_size_factors(counts(psce_herpud1[,psce_herpud1$condition == "ctrl"]), method = "ratio")))
glm_de_res2 <- lemur:::edger_test_de(glm_fit2, contrast = c(0, 0, 0, 0, 0, 0, 1))

plot(glm_de_res2$lfc, glm_de_res$lfc)
plot(glm_de_res2$pval, glm_de_res$pval, log = "xy", asp = 1); abline(0,1)

as_tibble(glm_de_res) %>%
  arrange(pval) %>%
  left_join(as_tibble(full_row_data), by = c("name" = "gid")) %>%
  print(n = 30)

```



```{r, paged.print=FALSE}
  ego1 <- clusterProfiler::enrichGO(gene = glm_de_res %>% slice_min(pval, n = 200) %>% pull(name), 
                                   universe = glm_de_res$name,
                                   OrgDb = org.Hs.eg.db::org.Hs.eg.db,
                                   keyType = "ENSEMBL",
                                   readable = TRUE,
                                   ont = "ALL")
  
  ego2 <- clusterProfiler::enrichGO(gene = glm_de_res %>% filter(lfc < -0.3) %>% slice_min(pval, n = 200) %>% pull(name), 
                                   universe = glm_de_res$name,
                                   OrgDb = org.Hs.eg.db::org.Hs.eg.db,
                                   keyType = "ENSEMBL",
                                   readable = TRUE,
                                   ont = "ALL")
  ekegg1 <- clusterProfiler::enrichKEGG(gene = glm_de_res %>% slice_min(pval, n = 200) %>% 
                                          pull(name) %>% 
                                          clusterProfiler::bitr(fromType = "ENSEMBL", toType = "ENTREZID", OrgDb = org.Hs.eg.db::org.Hs.eg.db, drop=TRUE) %>% 
                                          pull(ENTREZID), 
                                   universe = clusterProfiler::bitr(glm_de_res$name, fromType = "ENSEMBL", toType = "ENTREZID", OrgDb = org.Hs.eg.db::org.Hs.eg.db, drop=TRUE)$ENTREZID,
                                   organism = "hsa",
                                   keyType = "kegg")
  
  
  as_tibble(ego1) %>% print(n = 20)
  as_tibble(ego2) %>% print(n = 20)
  as_tibble(ekegg1) %>% print(n = 20)
  
  as_tibble(ego1) %>%
    filter(ID  == "GO:0034976")

  as_tibble(ego2) %>%
    filter(ID  == "GO:0006754")
```


```{r}
neg_log10_trans <- scales::trans_new("neg_log10", trans = \(x) -log10(x),  inv = \(x) 10^(-x), domain = c(1e-100, Inf))


volcano_enrichment_pl <- as_tibble(glm_de_res) %>%
  mutate(er_stress_related = name %in% ego1@geneSets[["GO:0034976"]]) %>%
  mutate(atp_syn_related = name %in% ego1@geneSets[["GO:0006754"]]) %>%
  mutate(category = case_when(
    er_stress_related ~ "ER Stress",
    atp_syn_related ~ "ATP biosynthesis",
    TRUE ~ "other"
  )) %>%
  arrange(desc(category)) %>%
  mutate(lfc = ifelse(lfc < -2.7, -Inf, lfc)) %>%
  ggplot(aes(x = lfc, y = pval)) +
    ggrastr::rasterize(geom_point(data = . %>% filter(category == "other"), color = "lightgrey", stroke = 0, size = 0.6), dpi = 300) +
    ggrastr::rasterize(geom_point(data = . %>% filter(category != "other"), aes(color = category), stroke = 0, size = 0.6), dpi = 300) +
    small_arrow(label = "up inside", position = c(0.7, 0.98), offset = unit(0.95, "npc"), label_offset = unit(-1.3, "mm"), fontsize = font_size_tiny) +
    small_arrow(label = "up outside",position = c(0.3, 0.02),  offset = unit(0.95, "npc"), label_offset = unit(-1.3, "mm"), fontsize = font_size_tiny) +
    geom_hline(yintercept = max(glm_de_res$pval[glm_de_res$adj_pval < 0.1])) +
    scale_color_manual(values = c("ER Stress" = "#023FA5", "ATP biosynthesis" = "#8E063B")) +
    scale_y_continuous(limits = c(1, NA), expand = expansion(mult = c(0, 0.05)), trans = neg_log10_trans,
                       breaks = 10^seq(0, -11, by = -2), labels = \(x) glue("$10^{{{log10(x)}}}$")) +
    scale_x_continuous(limits = c(-3, 3)) +
    guides(color = guide_legend(override.aes = list(size = 1))) +
    labs(x = "Log fold change", y = "p-value", color = "") +
    theme(legend.position = "bottom")


volcano_enrichment_pl
```





```{r, paged.print=FALSE}
dim(fit_small)
table(nei_pan_mod$adj_pval < 0.1)
nei_pan_tumor_small
table(fit_small$colData$cell_type)
sum(fit_small$colData$cell_type == "Tumor") - sum(filter(is_inside2, gene == "HERPUD1")$inside)
```



```{r}
cell_type_legend <- ggplot(enframe(cell_type_colors), aes(x = 0, y = 0, color = name)) + 
  geom_point() + 
  scale_color_manual(values = cell_type_colors, name = "Cell types") +
  theme(legend.position = "bottom")

condition_legend <- ggplot(enframe(condition_colors), aes(x = 0, y = 0, color = name)) + 
  geom_point() + 
  scale_color_manual(values = condition_colors, name = "Conditions") +
  theme(legend.position = "bottom")
```



```{r}
row1_umap_width <- 40
row1_umap_height <- 34
row1_offset_top <- 12

plot_assemble(
  add_text("(A) Glioblastoma experimental design", x = 1.5, y = 1.5, vjust = 1, fontsize = font_size, fontface = "bold"),
  add_graphic("../illustrations/glioblastoma_experiment_overview_wideArtboard 1.pdf", x = 1.5,  y = 2.5, units = "mm"),
  
  add_text("(B) Alignment with LEMUR adjusting for patient and treatment effect", x = 1.5, y = 24.5, vjust = 1, fontsize = font_size, fontface = "bold"),
  add_plot(cowplot::get_legend(condition_legend), x = 5, y = 28 , height = 3),
  add_text("UMAP of $Y$", x = 0 + row1_umap_width * 0.5, y = 32, vjust = 1, hjust = 0.5, fontsize = 7),
  add_text("UMAP of $Z$ before alignment", x = 0 + row1_umap_width * 1.5, y = 32, vjust = 1, hjust = 0.5, fontsize = 7),
  add_text("UMAP of $Z$ after alignment", x = 0 + row1_umap_width * 2.5, y = 32, vjust = 1, hjust = 0.5, fontsize = 7),
  add_plot(umap_orig_pc_pl, x = 0 + row1_umap_width * 0, y = 33, width = row1_umap_width, height = row1_umap_height),
  add_plot(umap_fit_pc_pl, x = 0 + row1_umap_width * 1, y = 33, width = row1_umap_width, height = row1_umap_height),
  add_plot(umap_fital_pc_pl, x = 0 + row1_umap_width * 2, y = 33, width = row1_umap_width, height = row1_umap_height),

  add_plot(cowplot::get_legend(cell_type_legend), x = 5, y = 67, height = 3),
  add_text("UMAP of $Y$", x = 0 + row1_umap_width * 0.5, y = 71, vjust = 1, hjust = 0.5, fontsize = 7),
  add_text("UMAP of $Z$ before alignment", x = 0 + row1_umap_width * 1.5, y = 71, vjust = 1, hjust = 0.5, fontsize = 7),
  add_text("UMAP of $Z$ after alignment", x = 0 + row1_umap_width * 2.5, y = 71, vjust = 1, hjust = 0.5, fontsize = 7),
  add_plot(umap_orig_ct_pl, x = 0 + row1_umap_width * 0, y = 72, width = row1_umap_width, height = row1_umap_height),
  add_plot(umap_fit_ct_pl, x = 0 + row1_umap_width * 1, y = 72, width = row1_umap_width, height = row1_umap_height),
  add_plot(umap_fital_ct_pl, x = 0 + row1_umap_width * 2, y = 72, width = row1_umap_width, height = row1_umap_height),

  add_text("(C) Cell type assignment", x = row1_umap_width * 3, y = 24.5, vjust = 1, fontsize = font_size, fontface = "bold"),
  add_plot(cell_type_dot_plot , x = 120, y = 28, width = 50, height = 38),
  add_text("(D) Multi-condition biplot", x = row1_umap_width * 3, y = 72, vjust = 1, fontsize = font_size, fontface = "bold"),
  add_plot(mc_biplot, x = row1_umap_width * 3 - 3, y = 75, width = 50, height = 33),
   
  add_text("(E) Cluster-free differential expression. Panobinostat causes", x = 1.5, y = 107, vjust = 1, fontsize = font_size, fontface = "bold"),
  add_graphic("../illustrations/manual_divergent_colorscaleArtboard 1.pdf", x = 82,  y = 105, units = "mm"),
  # add_plot(pan_volcano, x = 0, y = 111, width = 36, height = 38),
  add_plot(de_plots[[1]], x = 0, y = 110, width = 60, height = 36),
  add_plot(comparison_pl[[1]], x = 60, y = 118, width = 25, height = 28),
  add_plot(de_plots[[2]], x = 85, y = 110, width = 60, height = 36),
  add_plot(comparison_pl[[2]], x = 145, y = 118, width = 25, height = 28),

  add_text("(F) Panobinostat stimulates HERPUD1 up-regulation in a subpopulation of tumor cells", x = 1.5, y = 150, vjust = 1, fontsize = font_size, fontface = "bold"),
  add_plot(de_plots2, x = 0, y = 154, width = 61, height = 40),
  add_plot(comparison_pl2, x = 61, y = 154, width = 45, height = 40),
  add_plot(diff_pl, x = 104, y = 154, width = 25, height = 40),

  add_text("(G) Subpopulation markers\n\\quad\\quad in control condition", x = 130, y = 150, vjust = 1, fontsize = font_size, fontface = "bold"),
  add_plot(cowplot::get_legend(volcano_enrichment_pl + guides(color = guide_legend(override.aes = list(size = 1))) + labs(color = "")),
           x = 133, y = 156.5, width = 36, height = 4),
  add_plot(volcano_enrichment_pl + guides(color = "none"), x = 129, y = 160, width = 41, height = 37),
  
  width = 170, height = 197, units = "mm", show_grid_lines = FALSE, latex_support = TRUE,
  filename = "../plots/glioblastoma_results.pdf")
```



```{r, paged.print=FALSE}
tab <- colData(fit_small) %>%
  as_tibble() %>%
  mutate(treatment = factor(treatment, levels = c("vehicle (DMSO)", "0.2 uM panobinostat", "2.5 uM etoposide"))) %>%
    group_by(patient_id, treatment, pat_cond, age, gender, location) %>%
  summarize(n_cells = n(), .groups = "drop") %>%
  dplyr::select(`Patient ID` = patient_id, Age = age, Gender = gender,
                `Tumor location` = location, Conditon = treatment, `\\# Cells` = n_cells) %>%
  mutate(across(everything(), as.character)) %>%
  mutate(across(everything(), \(x) ifelse(x == lag(x, default = ""), "", x)), .by = `Patient ID`) %>%
  mutate(`Patient ID` = ifelse(`Patient ID` == lag(`Patient ID`, default = ""), "", `Patient ID`)) %>%
  kableExtra::kbl(booktabs = TRUE, format = "latex", escape = FALSE, linesep = "") %>%
  kableExtra::kable_styling(latex_options = "striped", stripe_index = c(1:2, 6:8, 12:14))

str_split(tab, "\n")[[1]] %>%
  magrittr::extract(., 2:(length(.)-1)) %>%
  write_lines("../plots/patient_overview_table.tex")
```





```{r}
sessionInfo()
```



